compiler/rustc_hir_typeck/src/method/suggest.rs

```diff
@@ -1363,7 +1363,111 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                    if ![
+                        lang.copy_trait(),
+                        lang.clone_trait(),
+                        tcx.get_diagnostic_item(sym::Debug),
+                        tcx.get_diagnostic_item(sym::PartialEq),
+                        tcx.get_diagnostic_item(sym::Default),
+                    ]
+                    .contains(&Some(trait_pred.def_id()))
+                    {
+                        // We restrict ourselves only to built-in `derive`s.
+                        continue;
+                    }
+                    let (adt, params) = match trait_pred.self_ty().kind() {
+                        ty::Adt(adt, params) if adt.did().is_local() => (*adt, params),
+                        _ => continue,
+                    };
+                    if tcx
+                        .all_impls(trait_pred.def_id())
+                        .filter_map(|imp_did| tcx.impl_trait_header(imp_did).map(|h| (imp_did, h)))
+                        .filter(|(did, header)| {
+                            let imp = header.trait_ref.instantiate_identity();
+                            let impl_adt = match imp.self_ty().ty_adt_def() {
+                                Some(impl_adt) if adt.did().is_local() => impl_adt,
+                                _ => return false,
+                            };
+                            header.polarity == ty::ImplPolarity::Positive
+                                && impl_adt == adt
+                                && tcx.is_automatically_derived(*did)
+                        })
+                        .count()
+                        == 1
+                    {
+                        // We now know that for this predicate, there *was* a `derive(Trait)` for
+                        // the trait at hand, so we don't want to suggest writing that again.
+                        for param in &params[..] {
+                            // Look at the type parameters for the currently obligated type to see
+                            // if a restriciton of `TypeParam: Trait` would help. If the
+                            // instantiated type param is not a type param but instead an actual
+                            // type, see if we can suggest `derive(Trait)` on *that* type.
+                            // See `tests/ui/suggestions/f1000.rs`
+                            let Some(ty) = param.as_type() else {
+                                continue;
+                            };
+                            match ty.kind() {
+                                ty::Adt(adt, _) if adt.did().is_local() => {
+                                    // The type param at hand is a local type, try to suggest
+                                    // `derive(Trait)`.
+                                    let trait_ref =
+                                        ty::TraitRef::identity(tcx, trait_pred.trait_ref.def_id)
+                                            .with_self_ty(tcx, ty);
```

Comment:
This isn't right :( `PartialEq` has two type parameters -- `Self` and `Rhs`. This only substitutes the first one.